vowpalwabbit/slates_label.cc

```diff
@@ -0,0 +1,208 @@
+  read_count += sizeof(TYPE);
+
+#define WRITE_CACHED_VALUE(VALUE, TYPE)           \
+  *(TYPE*)c = VALUE; \
+  c += sizeof(TYPE);
+
+size_t read_cached_label(shared_data*, void* v, io_buf& cache)
+{
+  // Since read_cached_features doesn't default the label we must do it here.
+  default_label(v);
+  slates::label* ld = static_cast<slates::label*>(v);
+
+  size_t read_count = 0;
+  char* read_ptr;
+  size_t next_read_size = 0;
+
+  READ_CACHED_VALUE(ld->type, slates::example_type);
+  READ_CACHED_VALUE(ld->weight, float);
+  READ_CACHED_VALUE(ld->labeled, bool);
+  READ_CACHED_VALUE(ld->cost, float);
+  READ_CACHED_VALUE(ld->slot_id, uint32_t);
+
+  uint32_t size_probs = 0;
+  READ_CACHED_VALUE(size_probs, uint32_t);
+
+  for (uint32_t i = 0; i < size_probs; i++)
+  {
+    ACTION_SCORE::action_score a_s;
+    READ_CACHED_VALUE(a_s, ACTION_SCORE::action_score);
+    ld->probabilities.push_back(a_s);
+  }
+  return read_count;
+}
+
+void cache_label(void* v, io_buf& cache)
+{
+  char* c;
+  slates::label* ld = static_cast<slates::label*>(v);
+  size_t size = sizeof(ld->type)
+      + sizeof(ld->weight)
+      + sizeof(ld->labeled)
+      + sizeof(ld->cost)
+      + sizeof(ld->slot_id)
+      + sizeof(uint32_t) // Size of probabilities
+      + sizeof(ACTION_SCORE::action_score) * ld->probabilities.size();
+
+  cache.buf_write(c, size);
+  WRITE_CACHED_VALUE(ld->type, slates::example_type);
+  WRITE_CACHED_VALUE(ld->weight, float);
+  WRITE_CACHED_VALUE(ld->labeled, bool);
+  WRITE_CACHED_VALUE(ld->cost, float);
+  WRITE_CACHED_VALUE(VW::convert(ld->slot_id), uint32_t);
+  WRITE_CACHED_VALUE(VW::convert(ld->probabilities.size()), uint32_t);
+  for (const auto& score : ld->probabilities)
+  {
+    WRITE_CACHED_VALUE(score, ACTION_SCORE::action_score);
+  }
+}
+
+float weight(void* v)
+{
+  return static_cast<polylabel*>(v)->slates.weight;
+}
+
+void default_label(void* v)
+{
+  auto& label = static_cast<polylabel*>(v)->slates;
+  label.type = example_type::unset;
+  label.weight = 1.f;
+  label.labeled = false;
+  label.cost = 0.f;
+  label.slot_id = 0;
+  label.probabilities.clear();
+}
+
+bool test_label(void* v)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  return ld.labeled == false;
+}
+
+void delete_label(void* v)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  ld.probabilities.delete_v();
+}
+
+void copy_label(void* dst, void* src)
+{
+  auto& ldDst = static_cast<polylabel*>(dst)->slates;
+  auto& ldSrc = static_cast<polylabel*>(src)->slates;
+
+  ldDst.type = ldSrc.type;
+  ldDst.weight = ldSrc.weight;
+  ldDst.labeled = ldSrc.labeled;
+  ldDst.cost = ldSrc.cost;
+  ldDst.slot_id = ldSrc.slot_id;
+  copy_array(ldDst.probabilities, ldSrc.probabilities);
+}
+
+void parse_label(parser* p, shared_data*, void* v, v_array<VW::string_view>& words)
+{
+  auto& ld = static_cast<polylabel*>(v)->slates;
+  ld.weight = 1;
+
+  if (words.size() < 2)
+    THROW("slates labels may not be empty");
```

Comment:
[Minor] Consider also adding something to explain the `words.size() == 1` case, as otherwise people who add "slates" but nothing else would get confused.